B::Keywords installation failure

macOS 11.4
Command:
> sudo cpan B::Keywords

Error msg:
t/11keywords.t .............. Can't open /System/Library/Perl/5.30/darwin-thread-multi-2level/CORE/keywords.h: No such file or directory at t/11keywords.t line 25.

Note:
Apple apparently moved these header files in an earlier release, and [this report](https://www.postgresql.org/message-id/153558865647.1483.573481613491501077%40wrigleys.postgresql.org) has some related information.

Full output:

Loading internal logger. Log::Log4perl recommended for better logging
Reading '/Users/[me]/.cpan/Metadata'
  Database was generated on Thu, 01 Jul 2021 00:56:02 GMT
Running install for module 'B::Keywords'
Checksum for /Users/[me]/.cpan/sources/authors/id/R/RU/RURBAN/B-Keywords-1.22.tar.gz ok
Configuring R/RU/RURBAN/B-Keywords-1.22.tar.gz with Makefile.PL
Checking if your kit is complete...
Looks good
Generating a Unix-style Makefile
Writing Makefile for B::Keywords
Writing MYMETA.yml and MYMETA.json
  RURBAN/B-Keywords-1.22.tar.gz
  /usr/bin/perl Makefile.PL -- OK
Running make for R/RU/RURBAN/B-Keywords-1.22.tar.gz
cp lib/B/Keywords.pm blib/lib/B/Keywords.pm
Manifying 1 pod document
  RURBAN/B-Keywords-1.22.tar.gz
  /usr/bin/make -- OK
Running make test for RURBAN/B-Keywords-1.22.tar.gz
PERL_DL_NONLAZY=1 "/usr/bin/perl" "-MExtUtils::Command::MM" "-MTest::Harness" "-e" "undef *Test::Harness::Switches; test_harness(0, 'blib/lib', 'blib/arch')" t/*.t
t/10basic.t ................. ok   
t/11keywords.t .............. Can't open /System/Library/Perl/5.30/darwin-thread-multi-2level/CORE/keywords.h: No such file or directory at t/11keywords.t line 25.
t/11keywords.t .............. skipped: (no reason given)
t/export.t .................. ok   
t/z_kwalitee.t .............. skipped: This test is only run for the module author
t/z_meta.t .................. skipped: This test is only run for the module author
t/z_perl_minimum_version.t .. skipped: Author tests not required for installation
t/z_pod-coverage.t .......... skipped: This test is only run for the module author
t/z_pod.t ................... ok   

Test Summary Report
- - - - - - - - - - - - - - - - - - -
t/11keywords.t            (Wstat: 512 Tests: 0 Failed: 0)
  Non-zero exit status: 2
Files=8, Tests=15,  1 wallclock secs ( 0.03 usr  0.01 sys +  0.35 cusr  0.08 csys =  0.47 CPU)
Result: FAIL
Failed 1/8 test programs. 0/15 subtests failed.
make: *** [test_dynamic] Error 255
  RURBAN/B-Keywords-1.22.tar.gz
  /usr/bin/make test -- NOT OK
//hint// to see the cpan-testers results for installing this module, try:
  reports RURBAN/B-Keywords-1.22.tar.gz

--- a/t/11keywords.t
+++ b/t/11keywords.t
@@ -26,19 +26,41 @@ sub _map_control_char {
     return $char;
 }
 
+sub kopen {
+    my @files = ();
+
+    # this is the default location for 'kewords.h', unless we're on macOS
+    push @files, File::Spec->catfile( $Config{archlibexp}, 'CORE', 'keywords.h' );
+
+    # Apple has restructured it's file system, if we're on macOS and the above
+    # doesn't exist, we're probably using system perl. try an alternate location
+    # via Apple's tools
+    if (($files[0] =~ /darwin/) and (! -e $files[0])) {
+	my $p = '';
+	$p = qx/xcrun --show-sdk-path/;
+	chomp $p;
+	push @files, File::Spec->catdir($p, $files[0]);
+    }
+
+    # return the first FH we can open from the list above. 99.44% of
+    # the time it's going to be the first (and correct) one
+    for my $file (@files) {
+	if (open my $fh, "<", $file) {
+	    return $fh;
+	}
+    }
+
+    # if we're here something else happend or apple is up to new tricks
+    die "Can't open any of @files";
+}
+
 # Test everything in keywords.h is covered.
 {
-    my $keywords = File::Spec->catfile( $Config{archlibexp}, 'CORE', 'keywords.h' );
-    if ($keywords =~ /darwin/) {
-        my $p = '';
-        $p = qx/xcrun --show-sdk-path/;
-        chomp $p;
-        $keywords = File::Spec->catdir($p, $keywords);
-    }
-    open FH, "< $keywords\0" or die "Can't open $keywords: $!";
+    # get a filehandle to 'keywords.h'
+    my $fh = kopen();
     local $/;
-    chomp( my @keywords = <FH> =~ /^\#define \s+ KEY_(\S+) /xmsg );
-    close FH;
+    chomp( my @keywords = <$fh> =~ /^\#define \s+ KEY_(\S+) /xmsg );
+    close $fh;
     my $usedevel = $Config{usedevel};
 
     my %covered = map { $_ => 1 } @Symbols, @Barewords;

--- a/t/11keywords.t
+++ b/t/11keywords.t
@@ -35,7 +35,7 @@ sub kopen {
     # Apple has restructured it's file system, if we're on macOS and the above
     # doesn't exist, we're probably using system perl. try an alternate location
     # via Apple's tools
-    if (($files[0] =~ /darwin/) and (! -e $files[0])) {
+    if ($^O eq "darwin" && $Config{archlibexp} =~ m!^/System/Library/Perl!) {
 	my $p = '';
 	$p = qx/xcrun --show-sdk-path/;
 	chomp $p;

--- a/t/11keywords.t
+++ b/t/11keywords.t
@@ -28,6 +28,12 @@ sub _map_control_char {
 # Test everything in keywords.h is covered.
 {
     my $keywords = File::Spec->catfile( $Config{archlibexp}, 'CORE', 'keywords.h' );
+    if ($keywords =~ /darwin/) {
+        my $p = '';
+        $p = qx/xcrun --show-sdk-path/;
+        chomp $p;
+        $keywords = File::Spec->catdir($p, $keywords);
+    }
     open FH, "< $keywords\0" or die "Can't open $keywords: $!";
     local $/;
     chomp( my @keywords = <FH> =~ /^\#define \s+ KEY_(\S+) /xmsg );

--- a/t/11keywords.t
+++ b/t/11keywords.t
@@ -1,5 +1,6 @@
-#!/usr/bin/perl -w
+# -*- perl -*-
 
+use warnings;
 use strict;
 use Test::More 'no_plan';
 

--- a/t/export.t
+++ b/t/export.t
@@ -1,5 +1,6 @@
-#!/usr/bin/perl -w
+# -*- perl -*-
 
+use warnings;
 use strict;
 use Test::More tests => 7;
 

